### PR TITLE
Feat/fail generating flow types for models with inconsistent naming of fields and getters

### PR DESCRIPTION
**Pull request by misino**
Each field in dto should contain getter with the corresponding name and there should be only one getter for a field.
It is not true, application throws an exception for incorrect DTO.

### CODE REVIEW

**Comment by havardh** [comment]
Hi @misino great contribution. 

I had a second opinion about how I would like this to be added and I was pleased to see that you have come half the way from my proposed solution to what I think is ideal. :)

I now propose that we hook this into the `verifier` phase instead of including it in the `parse` phase where I suggested and you've added it. Seem like the only thing you need to do is to move and rename `com.github.havardh.javaflow.phases.validator.ClassGetterNamingValidator` to `com.github.havardh.javaflow.phases.verifier.ClassGetterNamingVerifier`, and to make it implement `Verifier`. The latter would require you to make it have a non-static implementation and to accept a `List<Type>` inplace of a `Class`. As `Type` is the supertype of `Class` this should be quite simple, the other subtypes of `Type`can be bypassed inside of the `Verifier` by using `instanceof` ;) 

I really liked your idea of making this step report multiple violations instead of failing on the first problem. I think we would like a similar approach implemented for the `verifier`step as infrastructure implementation. It seems like the your current implementation will report only a single error per `Class`. We could make it report all errors it could find at once. Not sure how to best implement that though. But I think we need some Exception aggregation mechanism both inside the `Verifier` and outside. 

We could do something like:

```
class Execution {
 ...
  verify(types) {
    List<AggregatedException> exceptions
    for (Verifier verifier : verifiers) {
     for (Type type : types) {
        try {
          verifier.verify(type);
        } catch(AggregatedException e) {
           exceptions.push(e);
        }
      }
    }
    if (!exceptions.isEmpty()) {
      throw AggregatedExcepection(exceptions);
    }
  }
}

class VerifierOne extends implements {
  verify(Type type) {
     List<Exception> exceptions;
     for (Field field : type.fields() {
       if (field.isInvalid()) {
         expections.push(...);
       }
     }
  }
  if (!exceptions.isEmpty()) {
    throw AggregatedExcepection(exceptions);
  }
}
```

**Comment by misino** [comment]
I agree, it makes much more sense to verify in verification phase than in parse phase.

File: src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+
+import com.github.havardh.javaflow.ast.Class;
+import com.github.havardh.javaflow.ast.Field;
+import com.github.havardh.javaflow.ast.Method;
+import com.github.havardh.javaflow.ast.Type;
+import com.github.havardh.javaflow.exceptions.AggregatedException;
+import com.github.havardh.javaflow.exceptions.FieldGettersMismatchException;
+
+public class ClassGetterNamingVerifier implements Verifier {
```
Comment (havardh) [inline src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java]: This class is missing tests.

I wrote the attached proposal for unit tests for this class. **It is missing** one vital test which inspects the exception thrown by the verifier (namely the FieldGettersMismatchException). However there are two failing tests in this suite. Could you have a look at them and see if you agree on the proposed behaviour. 

1) I propose to make the verifier check all field in a class before reporting failures
2) I propose that we should also verify the type of the field and getter. 

```
diff --git a/src/main/java/com/github/havardh/javaflow/ast/builders/MethodBuilder.java b/src/main/java/com/github/havardh/javaflow/ast/builders/MethodBuilder.java
index c862aee..43cca49 100644
--- a/src/main/java/com/github/havardh/javaflow/ast/builders/MethodBuilder.java
+++ b/src/main/java/com/github/havardh/javaflow/ast/builders/MethodBuilder.java
@@ -1,4 +1,34 @@
 package com.github.havardh.javaflow.ast.builders;
 
-public class MethodBuilder {
+import com.github.havardh.javaflow.ast.Method;
+import com.github.havardh.javaflow.ast.Type;
+
+public class MethodBuilder implements Builder<Method> {
+
+  private String name;
+  private Type type;
+
+  public static MethodBuilder methodBuilder() {
+    return new MethodBuilder();
+  }
+
+  public MethodBuilder withName(String name) {
+    this.name = name;
+    return this;
+  }
+
+  public MethodBuilder withType(Type type) {
+    this.type = type;
+    return this;
+  }
+
+  /**
+   * Build a {@code Method} with the builder configuration
+   *
+   * @return the {@code Method}
+   */
+  @Override
+  public Method build() {
+    return new Method(name, type);
+  }
 }
diff --git a/src/test/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifierTest.java b/src/test/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifierTest.java
index d1a6edb..7bca225 100644
--- a/src/test/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifierTest.java
+++ b/src/test/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifierTest.java
@@ -1,5 +1,161 @@
-import static org.junit.jupiter.api.Assertions.*;
+package com.github.havardh.javaflow.phases.verifier;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.singletonList;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+import com.github.havardh.javaflow.ast.Class;
+import com.github.havardh.javaflow.ast.Type;
+import com.github.havardh.javaflow.ast.builders.ClassBuilder;
+import com.github.havardh.javaflow.ast.builders.FieldBuilder;
+import com.github.havardh.javaflow.ast.builders.MethodBuilder;
+import com.github.havardh.javaflow.ast.builders.TypeBuilder;
+import com.github.havardh.javaflow.exceptions.AggregatedException;
+import com.github.havardh.javaflow.model.CanonicalName;
 
 class ClassGetterNamingVerifierTest {
 
+  private static final Type PRIMITIVE_INT = TypeBuilder.primitive(CanonicalName.primitive("int"));
+  private static final Type PRIMITIVE_DOUBLE = TypeBuilder.primitive(CanonicalName.primitive("double"));
+
+  private ClassGetterNamingVerifier verifier = new ClassGetterNamingVerifier();
+
+  @Test
+  public void shouldPassForModelWithMatchingGetters() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("Class")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("someField")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .withGetter(MethodBuilder.methodBuilder()
+            .withName("getSomeField")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+
+    verifier.verify(singletonList(aClass));
+  }
+
+  @Test
+  public void shouldFailForModelWithGetterWithNonMatchingType() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("Class")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("someField")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .withGetter(MethodBuilder.methodBuilder()
+            .withName("getSomeField")
+            .withType(PRIMITIVE_DOUBLE)
+            .build())
+        .build();
+
+    AggregatedException exception = assertThrows(
+        AggregatedException.class,
+        () -> verifier.verify(singletonList(aClass))
+    );
+
+    assertThat(exception.getExceptions(), hasSize(1));
+  }
+
+  @Test
+  public void shouldFailForModelWithGetterWithNonMatchingName() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("Class")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("someField")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .withGetter(MethodBuilder.methodBuilder()
+            .withName("getSomeFields")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+
+    AggregatedException exception = assertThrows(
+        AggregatedException.class,
+        () -> verifier.verify(singletonList(aClass))
+    );
+
+    assertThat(exception.getExceptions(), hasSize(1));
+  }
+
+  @Test
+  public void shouldFailForModelWithoutGetters() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("Class")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("someField")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+
+    AggregatedException exception = assertThrows(
+        AggregatedException.class,
+        () -> verifier.verify(singletonList(aClass))
+    );
+
+    assertThat(exception.getExceptions(), hasSize(1));
+  }
+
+  @Test
+  public void shouldFailAndReportMultipleFieldsPerClass() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("Class")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("fieldOne")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("fieldTwo")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+
+    AggregatedException exception = assertThrows(
+        AggregatedException.class,
+        () -> verifier.verify(singletonList(aClass))
+    );
+
+    assertThat(exception.getExceptions(), hasSize(2));
+  }
+
+  @Test
+  public void shouldFailAndReportOnMultipleErroneousClasses() {
+    Class aClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("aClass")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("fieldOne")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+    Class bClass = ClassBuilder.classBuilder()
+        .withPackageName("com.github.havardh.models")
+        .withName("bClass")
+        .withField(FieldBuilder.fieldBuilder()
+            .withName("fieldOne")
+            .withType(PRIMITIVE_INT)
+            .build())
+        .build();
+
+    AggregatedException exception = assertThrows(
+        AggregatedException.class,
+        () -> verifier.verify(asList(aClass, bClass))
+    );
+
+    assertThat(exception.getExceptions(), hasSize(2));
+  }
+
}
```

File: src/main/java/com/github/havardh/javaflow/exceptions/VerifierException.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package com.github.havardh.javaflow.exceptions;
```
Comment (havardh) [inline src/main/java/com/github/havardh/javaflow/exceptions/VerifierException.java]: This doesn't seem to be in use

File: src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        try {
+          validate((Class) type);
+        } catch (Exception e) {
+          exceptions.add(e);
+        }
+      }
+    }
+
+    if (!exceptions.isEmpty()) {
+      throw new AggregatedException("Class getter naming validation failed", exceptions);
```
Comment (havardh) [inline src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java]: We need to add a space after this message, either here or in the AggregatedException class.

File: src/main/java/com/github/havardh/javaflow/exceptions/AggregatedException.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.github.havardh.javaflow.exceptions;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class AggregatedException extends RuntimeException {
```
Comment (havardh) [inline src/main/java/com/github/havardh/javaflow/exceptions/AggregatedException.java]: Would be nice with some unit tests on this class to verify the message output when the exception is thrown.

File: src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java (hunk 1 above)
Comment (misino) [reply src/main/java/com/github/havardh/javaflow/phases/verifier/ClassGetterNamingVerifier.java]: I've updated tests and added a validation for compatibility of field and getter type.

**Pull request merged**